monitor.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2023 The Ebitengine Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebiten
+
+import (
+	"github.com/hajimehoshi/ebiten/v2/internal/ui"
+)
+
+// Monitor represents a monitor available to the system.
+type Monitor = ui.Monitor
+
+// WindowMonitor returns the current monitor. If a window has not been created, this will be nil.
+func WindowMonitor() *Monitor {
+	return ui.Get().Monitor()
+}
+
+// SetWindowMonitor sets the monitor that the window should be on. This can be called before or after Run.
+func SetWindowMonitor(monitor *Monitor) {
+	ui.Get().Window().SetMonitor(monitor.ID())
+}
+
+// SetWindowMonitorByID sets the monitor by its ID that the window should be on. This can be called before or after run.
+func SetWindowMonitorByID(monitorID int) {
```

Comment:
IDs are returned by the Monitor struct's ID() function. This is necessary for storing and restoring target monitor in game configuration files.

Reply:
But do we need `SetWindowMonitorByID`? Isn't `SetWindowMonitor` enough?

Reply:
> This is necessary for storing and restoring target monitor in game configuration files.

I guess IDs can change by chaning the monitor setups, so storing IDs would not make sense...

Reply:
To clarify, IDs are mapped directly to the slice index returned by `glfw.GetMonitors()`. These indexes always have 0 as the primary monitor and will always be in the same order between system restarts. By index is the only way glfw can provide any form of monitor identification, unfortunately.

If `SetWindowMonitorByID` would be removed, the Monitor struct would have to have an `ID` property instead of a function that the end-user could then restore the monitor by issuing: `SetWindowMonitor(Monitor{ID: myId})`. This feels a little strange.

In some ways it would make more sense to not use a Monitor struct, but rather just use functions that take monitor IDs, working in a similar way to TouchIDs. However, this would add lots of extra functions to the ebiten package.

Reply:
So IDs except for 0 doesn't seem so reliable to me: when relaunching the process, there is no guarantee that the same ID is available, right?

Reply:
According to a glfw developer on the glfw forum, the order of monitors returned by glfw is guaranteed, so therefore the same ID will always match the same monitor. I've tested it myself and it is reliable on both Linux and Windows, between process relaunching and system rebooting.

Reply:
I see.

If you change the monitor setups, the relashipships of IDs and monitors can be changed, right? I was wondering what you meant by storing IDs. Is this storing IDs into a file? What is an actual use case?

Reply:
Correct, if you unplug/re-arrange monitors the order will change. If the monitors end up in the same configuration, their index will always be the same, so it doesn't get "randomized", so to speak.

There are two use cases:

  * 1. Video Game Options
    * User wants to start their game on a specific monitor, so in an options menu they choose a monitor to use (0,1,2, or some nicer names).
    * This data must be saved, so presume a JSON file. The only identifiable information to store about a monitor is its ID (or its index, to be specific). So that gets marshaled into the configuration file.
    * On next start, the configuration file is read and the monitor's ID is used on start using a call like `SetWindowMonitorById(config.MonitorID)`
    * User is happy the window is on the correct monitor and that the video game remembers it.

  * 2. Developer ease of use
    * While developing with automatic rebuilding of an ebiten program, the window spawns in front of the IDE as its the active monitor.
    * To remedy this, the developer adds a `--monitor=1` flag to their program so that the window spawns on their second monitor
    * Developer is happy and doesn't have to use auto hotkeys or other macro systems to position their window.

Hopefully this was clear enough. I plan to use a monitor target option in a commercial video game I am working on at the moment. There is a menu option for it, but nothing to put there yet. For development, it would help me a lot as well to be able to target the monitor to use.

Reply:
In order to identify a monitor, I think we can use a left-upper position of a monitor, for example. Even if the user modifies the monitors, an app can choose a monitor which is close to the position (or you can just fallback to the primary monitor). In order to restore the state, remembering a position seems to make more sense.

If you want to use integer IDs, assuming IDs are just indices of a monitor slice, we can do `SetWindowMonitor(monitors[id])` instead of `SetWindowMonitorById(id)`. Am I missing something?

Reply:
I would have to go through my games to determine which ones have the feature, but a few from my steam are:

  * Doom 2016
  * Noita
  * Factorio
  * Streets of Rogue

These refer to the monitors as either monitors or screens and present them as integers in a drop-down list. Most of them started at 1. It is certainly less common to present this option, but I have appreciated it when it is available.

In terms of identifying a monitor using positions, how would that be presented to the end-user? Would we iterate through the monitors, get their position, then present that to the user in a way such as: `0x0`, `1920x0`, `3840x0`? That seems less friendly than 0, 1, 2, but it could work. If this information wasn't presented to the user at all, but internally managed in the program by storing position and restoring it, it would work fine.

I think most of the time if the end-user is modifying their monitor layout, any presumption of which monitor is the right one is lost. The only way to reliably identify monitors would be to use a hardware identifier, but that is unfortunately unavailable through glfw.

If we use `monitors[id]`, the developer would have to check for validity first, as if `monitors` was reduced in length due to unplugging a monitor or two, but the developer passed in `id`, that would be out of bounds. This could work.

Reply:
Thank you for confirming existing games.

You can still show indices to users and store them. It is up to you. But even in this case, we don't need `SetWindowMonitorById`. As you pointed out, we have to check the ID is in the range, but this should not be so cumbersome.

Reply:
Okay, I will rewrite to use `AppendMonitors` and get rid of `SetWindowMonitorById`. I think the Monitor struct is still useful, as it gives Bounds and Name on platforms that supports monitor names.

The following should be the end-result.

```golang
if monitors := ebiten.AppendMonitors(nil); len(monitors) > id {
  ebiten.SetWindowMonitor(Monitor{ID: config.MonitorID})
}
```

Reply:
> ebiten.SetWindowMonitor(Monitor{ID: config.MonitorID})

Making a Monitor struct from Ebitengine users doesn't make sense. Why not just `monitors[config.MonitorID]`?